Automattic/isolated-block-editor
# Collab: Sanitize HTML received from peers

This sanitizes any HTML received from peers to prevent XSS attacks.

## Background

There are ways you can "self-XSS" yourself in the block editor, because certain blocks allow the user to inject arbitrary HTML into the DOM. This is not a problem in normal editor usage, because any arbitrary HTML written by other users will be sanitized server-side (`wp_kses`) before being loaded into the editor.

In collab mode however, this kind of arbitrary HTML can be instantly synced over from remote peers without server-side sanitization.

## Testing instructions

1. `yarn storybook`
2. Open two browser tabs at http://localhost:6006/?path=/story/collaboration--default
3. In one tab, insert a Custom HTML block. Try to add some JS, such as in a `<script>` tag or an `onclick` attribute.
4. Safe HTML should sync over to the other tab as expected, but any unsafe scripts should be stripped out.
5. No scripts should run when you click the Preview button in the Custom HTML block.

    <img src="https://user-images.githubusercontent.com/555336/151040116-5dfc8ac4-8ffc-4eb4-a7e9-04714e1da138.png" alt="Preview button in Custom HTML block toolbar" width="300">

### Automated testing

✅ Unit tests for the sanitization function
✅ UI integration tests to smoke test that it's working
